**Following guidelines set expectations for participant behaviour during workshop activities. They also ensure that the class environment is welcoming, inclusive, and respectful.**
- Where a discussion is taking place, allow everyone a chance to speak
- Listen respectfully, without interrupting and with an open mind to understanding others’ views
- Be professional and productive, and always share your ideas, your opinion matters, we all can learn something from each other
- Personal information that comes up in the conversation should be kept confidential
- Avoid inflammatory language
- Avoid assumptions about any member of the class or generalisations about social groups

**You have now joined a group of fellow analysts in a workshop:**
#### Outcomes:

- To be able to answer the questions to test your knowledge of File Operations and Regular Expressions


#### Note:
- We understand learners will progress at their own speed
- Tutors will be on hand to answer questions

# 🚀 File Operations and Regular Expressions Practical

# ☑️ Part 1: File operations

You now understand that Python treats files as objects that must be **opened** before they can be **read from** or **written to**.

Next, we’ll apply this knowledge in a practical context by working with actual files.

The goal is to load external data into Python and store it using familiar data structures like **lists** and **dictionaries**. Once loaded, you’ll be able to manipulate, analyse, or transform the data however you need.

## ✅ Question 1: Opening a file and reading it

In the `data` folder is a file named `poem.txt`.

As part of a `with`... `as` statement, use the `open()` function to access the file and store it as a variable named `poem`.

Then, look at what data type the variable is and how long it is, use `type()` and `len()` functions.

See below code syntax for some guidance:
```python
with open('data/<file_name.txt>', 'r') as f:
    poem = f.read()

print(type(poem))
print(len(poem))
```

In [10]:
#Add your code below

with open('data/poem.txt', 'r') as f:
    poem = f.read()

print(type(poem))
print(len(poem))



<class 'str'>
611


*__Windows users:__ If you get an error, you may need to use `'rb'` ('read binary') as the second argument when using the `open()` function to read files; if so, this applies to all questions involving reading files.*

There are two ways to inspect a variable in a notebook. You can just type the variable name, to see the raw data:

In [11]:
poem

"Three witches, casting a spell ...\n\nRound about the cauldron go;\nIn the poison'd entrails throw.\nToad, that under cold stone\nDays and nights hast thirty one\nSwelter'd venom sleeping got,\nBoil thou first i' the charmed pot.\n\nDouble, double toil and trouble;\nFire burn and cauldron bubble.\n\nFillet of a fenny snake,\nIn the cauldron boil and bake;\nEye of newt, and toe of frog,\nWool of bat, and tongue of dog,\nAdder's fork, and blind-worm's sting,\nLizard's leg, and howlet's wing,\nFor a charm of powerful trouble,\nLike a hell-broth boil and bubble.\n\nDouble, double toil and trouble;\nFire burn and cauldron bubble.\n"

Or you can use `print()`, which will format it using any special characters specified within the variable, like tabs or linebreaks:

In [12]:
print(poem)

Three witches, casting a spell ...

Round about the cauldron go;
In the poison'd entrails throw.
Toad, that under cold stone
Days and nights hast thirty one
Swelter'd venom sleeping got,
Boil thou first i' the charmed pot.

Double, double toil and trouble;
Fire burn and cauldron bubble.

Fillet of a fenny snake,
In the cauldron boil and bake;
Eye of newt, and toe of frog,
Wool of bat, and tongue of dog,
Adder's fork, and blind-worm's sting,
Lizard's leg, and howlet's wing,
For a charm of powerful trouble,
Like a hell-broth boil and bubble.

Double, double toil and trouble;
Fire burn and cauldron bubble.



Linebreaks are represented by the newline character, `\n`. This may look like two separate characters, but Python treats it as one!

Currently, our data is one long string. If we want each line of the poem to be its own string, there are two ways to do this.

Use `readlines()` instead of `read()`to return a list of strings (the newline character `\n` will determine where each string ends):

In [13]:
with open('data/poem.txt', 'r') as f:
    poem = f.readlines()
    
poem

['Three witches, casting a spell ...\n',
 '\n',
 'Round about the cauldron go;\n',
 "In the poison'd entrails throw.\n",
 'Toad, that under cold stone\n',
 'Days and nights hast thirty one\n',
 "Swelter'd venom sleeping got,\n",
 "Boil thou first i' the charmed pot.\n",
 '\n',
 'Double, double toil and trouble;\n',
 'Fire burn and cauldron bubble.\n',
 '\n',
 'Fillet of a fenny snake,\n',
 'In the cauldron boil and bake;\n',
 'Eye of newt, and toe of frog,\n',
 'Wool of bat, and tongue of dog,\n',
 "Adder's fork, and blind-worm's sting,\n",
 "Lizard's leg, and howlet's wing,\n",
 'For a charm of powerful trouble,\n',
 'Like a hell-broth boil and bubble.\n',
 '\n',
 'Double, double toil and trouble;\n',
 'Fire burn and cauldron bubble.\n']

Second, we could process the data _after_ loading it. Since it is a `str`, it will have a `.splitlines()` method that can be used to chop it up into a list of smaller strings; use this method on `poem`:

In [14]:
with open('data/poem.txt', 'r') as f:
    poem = f.read()

poem = poem.splitlines()
poem

['Three witches, casting a spell ...',
 '',
 'Round about the cauldron go;',
 "In the poison'd entrails throw.",
 'Toad, that under cold stone',
 'Days and nights hast thirty one',
 "Swelter'd venom sleeping got,",
 "Boil thou first i' the charmed pot.",
 '',
 'Double, double toil and trouble;',
 'Fire burn and cauldron bubble.',
 '',
 'Fillet of a fenny snake,',
 'In the cauldron boil and bake;',
 'Eye of newt, and toe of frog,',
 'Wool of bat, and tongue of dog,',
 "Adder's fork, and blind-worm's sting,",
 "Lizard's leg, and howlet's wing,",
 'For a charm of powerful trouble,',
 'Like a hell-broth boil and bubble.',
 '',
 'Double, double toil and trouble;',
 'Fire burn and cauldron bubble.']

## ✅ Question 2: Opening docx files

Import the package `docx2txt` and `.process()` the file at `data/air_quality.docx`, assigning the result to `doc_text`.

Check the `type()` of `doc_text`.


See below code syntax for some guidance:
```python
import docx2txt
doc_text = docx2txt.process('data/<file_name>.docx')
```

In [15]:
import docx2txt

#Add your code below
#doc_text = ...

doc_text = docx2txt.process('data/air_quality.docx')
print(doc_text)
type(doc_text)


Air Quality Steering Group – Meeting 19



Thursday 19th September 2019



Present:

Sandra Coltman (Sec)	(SC)		WCC Environmental Health and Licensing

Councillor L Murphy		(LM)		Portfolio Holder for Environment

Councillor F Pearson	(FP)		Shadow Portfolio Holder

Sara Davis			(SD)		Winchester BID

Phil Gagg			(PG)		Win Acc	

David Ingram (Chair)	(DI)		WCC Service Lead – Public Protection

Lucy McKeown		(LMc)		WCC Transport Planner

Phil Tidridge                        (PT)              WCC Scientific Officer



Apologies for Absence 



Apologies for absence were submitted on behalf of Simon Finch (SF) WCC Corporate Head of Regulatory, Richard Hein (RH) WCC Head of Parking Services and CCTV, Andy Hickman (AH) Head of Programme, James Moore (JM) Transport, Hampshire County Council, Dan Massey (DM) WCC Engineering and Transport, Mike Slinn (MS)  TSP Transport Group Paul Spencer (PS) Executive Director, Winchester Business Improvement District,  Darren Carmichael (DC) HCC and Richard Bot

str

Now that you have a string, use the `.splitlines()` method to return a list of the contents of each line, assigning the result to `lines`:

In [16]:
lines = doc_text.splitlines()
lines

['Air Quality Steering Group – Meeting 19',
 '',
 '',
 '',
 'Thursday 19th September 2019',
 '',
 '',
 '',
 'Present:',
 '',
 'Sandra Coltman (Sec)\t(SC)\t\tWCC Environmental Health and Licensing',
 '',
 'Councillor L Murphy\t\t(LM)\t\tPortfolio Holder for Environment',
 '',
 'Councillor F Pearson\t(FP)\t\tShadow Portfolio Holder',
 '',
 'Sara Davis\t\t\t(SD)\t\tWinchester BID',
 '',
 'Phil Gagg\t\t\t(PG)\t\tWin Acc\t',
 '',
 'David Ingram (Chair)\t(DI)\t\tWCC Service Lead – Public Protection',
 '',
 'Lucy McKeown\t\t(LMc)\t\tWCC Transport Planner',
 '',
 'Phil Tidridge                        (PT)              WCC Scientific Officer',
 '',
 '',
 '',
 'Apologies for Absence ',
 '',
 '',
 '',
 'Apologies for absence were submitted on behalf of Simon Finch (SF) WCC Corporate Head of Regulatory, Richard Hein (RH) WCC Head of Parking Services and CCTV, Andy Hickman (AH) Head of Programme, James Moore (JM) Transport, Hampshire County Council, Dan Massey (DM) WCC Engineering and Transport, Mi

Remove all blank lines (i.e. empty strings) from the list of `lines`, assigning the remaining list of lines to `content`:

- Don't forget that [Stack Overflow](https://stackoverflow.com/questions/3845423/remove-empty-strings-from-a-list-of-strings) can help you!

In [17]:
content = []

for line in lines:
    if line  !=  '':
        content.append(line)

content     

['Air Quality Steering Group – Meeting 19',
 'Thursday 19th September 2019',
 'Present:',
 'Sandra Coltman (Sec)\t(SC)\t\tWCC Environmental Health and Licensing',
 'Councillor L Murphy\t\t(LM)\t\tPortfolio Holder for Environment',
 'Councillor F Pearson\t(FP)\t\tShadow Portfolio Holder',
 'Sara Davis\t\t\t(SD)\t\tWinchester BID',
 'Phil Gagg\t\t\t(PG)\t\tWin Acc\t',
 'David Ingram (Chair)\t(DI)\t\tWCC Service Lead – Public Protection',
 'Lucy McKeown\t\t(LMc)\t\tWCC Transport Planner',
 'Phil Tidridge                        (PT)              WCC Scientific Officer',
 'Apologies for Absence ',
 'Apologies for absence were submitted on behalf of Simon Finch (SF) WCC Corporate Head of Regulatory, Richard Hein (RH) WCC Head of Parking Services and CCTV, Andy Hickman (AH) Head of Programme, James Moore (JM) Transport, Hampshire County Council, Dan Massey (DM) WCC Engineering and Transport, Mike Slinn (MS)  TSP Transport Group Paul Spencer (PS) Executive Director, Winchester Business Improve

# ☑️ Part 2: Regular Expressions

Look at the file `data/nasa_web.log`. It contains two month's worth of all HTTP requests to the NASA Kennedy Space Center server in Florida. 
We are going to go though that file and extract some interesting information. 

Let's start by reading the file into your python session.

In [18]:
with open('data/nasa_web.log', 'r') as f:
    log_file = f.read()

log_file

'199.72.81.55 - - [01/Jul/1995:00:00:01 -0400] "GET /history/apollo/ HTTP/1.0" 200 6245\nunicomp6.unicomp.net - - [01/Jul/1995:00:00:06 -0400] "GET /shuttle/countdown/ HTTP/1.0" 200 3985\n199.120.110.21 - - [01/Jul/1995:00:00:09 -0400] "GET /shuttle/missions/sts-73/mission-sts-73.html HTTP/1.0" 200 4085\nburger.letters.com - - [01/Jul/1995:00:00:11 -0400] "GET /shuttle/countdown/liftoff.html HTTP/1.0" 304 0\n199.120.110.21 - - [01/Jul/1995:00:00:11 -0400] "GET /shuttle/missions/sts-73/sts-73-patch-small.gif HTTP/1.0" 200 4179\nburger.letters.com - - [01/Jul/1995:00:00:12 -0400] "GET /images/NASA-logosmall.gif HTTP/1.0" 304 0\nburger.letters.com - - [01/Jul/1995:00:00:12 -0400] "GET /shuttle/countdown/video/livevideo.gif HTTP/1.0" 200 0\n205.212.115.106 - - [01/Jul/1995:00:00:12 -0400] "GET /shuttle/countdown/countdown.html HTTP/1.0" 200 3985\nd104.aa.net - - [01/Jul/1995:00:00:13 -0400] "GET /shuttle/countdown/ HTTP/1.0" 200 3985\n129.94.144.152 - - [01/Jul/1995:00:00:13 -0400] "GET / 

If you're not familiar with HTTP logs, the data provided can be broken down like this:

`netspace.net.au - - [01/Jul/1995:03:34:44 -0400] "GET /shuttle/countdown/ HTTP/1.0" 200 3985`



| Host making the request | Timestamp            | Timezone      | HTTP method | URL                 | HTTP Version |HTTP status code | Bytes in the reply |
|-------------------------|----------------------|---------------|-------------|---------------------|-------------|-----------------|--------------------|
| netspace.net.au         | 01/Jul/1995:03:34:44 | -0400         | GET         | /shuttle/countdown/ | HTTP/1.0    | 200              | 3985               |

## The vocabulary

Regular expressions are a way to define strings that match a specific pattern. There is a whole language to be learned here and below you will find some of the most useful values.

    \d      Any Digit
    \D      Any Non-digit character
    .       Any Character
    \.      Period
    [abc]   Only a, b, or c
    [^abc]  Not a, b, nor c
    [a-z]   Characters a to z
    [0-9]   Numbers 0 to 9
    \w      Any Alphanumeric character
    \W      Any Non-alphanumeric character
    {m}     m Repetitions
    {m,n}   m to n Repetitions
    *       Zero or more repetitions
    +       One or more repetitions
    ?       Optional character
    \s      Any Whitespace
    \S      Any Non-whitespace character
    ^…$     Starts and ends
    (ab|de) Matches ab or de

[source](https://regexone.com/references/python)

A fantastic way of learning/practicing regex is [regex101](https://regex101.com/)

# Match patterns

Firstly, try to understand the regular expression below:

In [19]:
r'GET /\w+'

'GET /\\w+'

This regular expression pattern looks for the word **`GET`**, followed by a **space character** ` `, a **forward slash** `/`, and then one or more **alphanumeric characters** (letters, digits, or underscore). This pattern is commonly used to search for **simple URLs** requested with a `GET` method in logs or text.

Although not always required, it's a good practice to prefix regex patterns with **`r''`** (raw strings). This ensures that backslashes `\` are treated **literally** and not as Python escape characters. For example, `\n` is a newline in a normal string but just a backslash and "n" in a raw string.

# Match dates

Another example of a simple regular expression would be to search for a date. Imagine we wanted to capture all date instances that followed the convention 01/Jul/1995. This pattern would look like:

In [20]:
r'\d{2}/[a-zA-Z]{3}/\d{4}'

'\\d{2}/[a-zA-Z]{3}/\\d{4}'

Pattern Breakdown:
- `\d{2}` — Matches **exactly two digits**, e.g. `01`, `23`  
- `/` — Matches a **literal forward slash**  
- `[a-zA-Z]{3}` — Matches **exactly three letters**, either uppercase or lowercase, e.g. `Jul`, `DEC`  
- `/` — Matches another **literal forward slash**  
- `\d{4}` — Matches **exactly four digits**, e.g. `1995`, `2023`

With our log we know exactly the number of letters and numbers expected. With regex, you can use `{n}` after a pattern to specify that you expect **n** instances. For example `\d{4}` will match exactly 4 digits.

## ✅ Question 3: `search()`

Python supports regular expressions through a built-in library called `re`. Start by importing the `re` library and using 
the `.search()` function, which takes a regex pattern and a string to be searched. 

- Search `log_file` for `GET` requests (the regex specified under the vocublary list above can be used) `r'GET /\w+'`

- then assign the `re.Match object` which is returned by `re.search()`, to the variable `match`

See below code syntax for some guidance:

```python
match = re.search(<regex_pattern>, <data>)
```

In [21]:
import re

# Add your code below
pattern = r'GET /\w+'
match = re.search(pattern, log_file)
match


<re.Match object; span=(47, 59), match='GET /history'>

Note that the search function will only return the first match, so it is mostly useful to see whether your regular expression does at least have one hit. If there is no hit, it simply returns `None`. If it does find a hit, it returns the hit as a `MatchObject`.

We can use the `.group()` method of the `MatchObject` to extract the actual string that was matched:

In [22]:
match.group()

'GET /history'

We can use the `.start()` method of the `MatchObject` to extract the starting position of the string that was matched:

In [23]:
match.start()

47

We can use the `.end()` method of the `MatchObject` to extract the end position of the string that was matched:

In [24]:
match.end()

59

### Creating a message from `MatchObject` properties

Similarly to the `.group()` method which extracts the actual string which was matched, you can use the `.start()` and `.end()` methods to collect the position of where the regular expression matched.

`print` the `start` and `end` positions formatted nicely with the actual match. Use an f-string to put together the `.group()`, `.start()`, `.end()` into a sentence as a single string, assigned to `message`:

In [25]:
# return a single string containing .group(), .start() and .end() of the match
message = f'Found "{match.group()}" at position: {match.start()}-{match.end()}'
message


'Found "GET /history" at position: 47-59'

## ✅ Question 4: Finding and Replacing IP Addresses

Your task is to **identify and replace all IPv4 host IP addresses** in the `log_file` with the string **`'unknown domain'`**.

This task will be completed in **two steps**:

1. **Identify** all IP addresses in the file  
2. **Replace** each IP address with `'unknown domain'`

To help with this, you've been provided the following regular expression pattern:

```python
r'[0-9]*\.[0-9]*\.[0-9]*\.[0-9]*'
```

This pattern is designed to match IPv4 addresses such as `192.168.0.1`. Here's a quick breakdown of how it works:

- `[0-9]*` – Matches **zero or more digits**, representing each segment of the IP address  
- `\.` – Matches a **literal dot** (`.`), which separates each segment  
- The pattern is repeated **four times** to match all four parts of a typical IPv4 address

---

#### Step 1: Finding IP Addresses

Now find all IP addresses in the `log_file`, you can use the `re.findall()` function.

See below code syntax for some guidance:

```python
pattern = r'[0-9]*\.[0-9]*\.[0-9]*\.[0-9]*'

re.findall(<regex_pattern>, <data>)
```

In [26]:
# Step 1: Finding IP Addresses
# Add your code below

re.findall(pattern, log_file)


['GET /history',
 'GET /shuttle',
 'GET /shuttle',
 'GET /shuttle',
 'GET /shuttle',
 'GET /images',
 'GET /shuttle',
 'GET /shuttle',
 'GET /shuttle',
 'GET /shuttle',
 'GET /images',
 'GET /images',
 'GET /shuttle',
 'GET /images',
 'GET /images',
 'GET /images',
 'GET /images',
 'GET /facts',
 'GET /shuttle',
 'GET /images',
 'GET /shuttle',
 'GET /shuttle',
 'GET /shuttle',
 'GET /shuttle',
 'GET /images',
 'GET /shuttle',
 'GET /history',
 'GET /images',
 'GET /shuttle',
 'GET /images',
 'GET /images',
 'GET /images',
 'GET /images',
 'GET /images',
 'GET /shuttle',
 'GET /shuttle',
 'GET /shuttle',
 'GET /shuttle',
 'GET /shuttle',
 'GET /history',
 'GET /software',
 'GET /software',
 'GET /images',
 'GET /software',
 'GET /shuttle',
 'GET /cgi',
 'GET /shuttle',
 'GET /shuttle',
 'GET /news',
 'GET /shuttle',
 'GET /shuttle',
 'GET /shuttle',
 'GET /software',
 'GET /images',
 'GET /history',
 'GET /shuttle',
 'GET /shuttle',
 'GET /history',
 'GET /shuttle',
 'GET /shuttle',
 '

#### Step 2: Replacing IP Addresses and Saving the File

Now that we've identified the IP addresses, let's move on to replacing them using `re.sub()` function.

We'll use the same regular expression pattern to substitute every IPv4 address in the `log_file` with the string **`'unknown domain'`**.

Finally, we'll **save the updated content** to a new file called `log_file_edited.log`

In [27]:
# Step 2: Replacing IP Addresses and Saving the File

pattern = r'[0-9]*\.[0-9]*\.[0-9]*\.[0-9]*'
replacement = r'unknown domain'

with open('data/log_file_edited.log', 'w') as f:
    f.write(re.sub(pattern, replacement, log_file))

In [28]:
# Read and Print the Edited File for verification

with open('data/log_file_edited.log', 'r') as f:
    content = f.read()
    print(content)

unknown domain - - [01/Jul/1995:00:00:01 -0400] "GET /history/apollo/ HTTP/1.0" 200 6245
unicomp6.unicomp.net - - [01/Jul/1995:00:00:06 -0400] "GET /shuttle/countdown/ HTTP/1.0" 200 3985
unknown domain - - [01/Jul/1995:00:00:09 -0400] "GET /shuttle/missions/sts-73/mission-sts-73.html HTTP/1.0" 200 4085
burger.letters.com - - [01/Jul/1995:00:00:11 -0400] "GET /shuttle/countdown/liftoff.html HTTP/1.0" 304 0
unknown domain - - [01/Jul/1995:00:00:11 -0400] "GET /shuttle/missions/sts-73/sts-73-patch-small.gif HTTP/1.0" 200 4179
burger.letters.com - - [01/Jul/1995:00:00:12 -0400] "GET /images/NASA-logosmall.gif HTTP/1.0" 304 0
burger.letters.com - - [01/Jul/1995:00:00:12 -0400] "GET /shuttle/countdown/video/livevideo.gif HTTP/1.0" 200 0
unknown domain - - [01/Jul/1995:00:00:12 -0400] "GET /shuttle/countdown/countdown.html HTTP/1.0" 200 3985
d104.aa.net - - [01/Jul/1995:00:00:13 -0400] "GET /shuttle/countdown/ HTTP/1.0" 200 3985
unknown domain - - [01/Jul/1995:00:00:13 -0400] "GET / HTTP/1.0"